### PR TITLE
OCPBUGS-33789: [Nutanix] installer intermittent failed at uploading the bootstrap image to PC when CAPI enabled

### PR DESCRIPTION
**Pull request by yanhua121**
OCPBUGS-33789
[Nutanix] installer intermittent failed to upload bootstrap image data to prism-central when CAPI enabled

My investigation shows this intermittent image uploading issue is at the prism-api server side with some older version of the prism-central server. Since the issue is intermittent, we can workaround it by retrying a number of times if uploading image fails.

### CODE REVIEW

File: pkg/infrastructure/nutanix/clusterapi/clusterapi.go
Comment (r4f4) [inline pkg/infrastructure/nutanix/clusterapi/clusterapi.go:261]:
```suggestion
		return fmt.Errorf("failed to delete the image with uuid %s: %w", imgUUID, err)
```

File: pkg/infrastructure/nutanix/clusterapi/clusterapi.go
Comment (r4f4) [inline pkg/infrastructure/nutanix/clusterapi/clusterapi.go:273]:
```suggestion
	}
	if taskUUID, ok := respd.Status.ExecutionContext.TaskUUID.(string); ok {
		logrus.Infof("deleting the image with uuid %s, taskUUID: %s", imgUUID, taskUUID)

		// Wait till the image creation task is successed.
		if err := nutanixtypes.WaitForTask(nutanixCl.V3, taskUUID); err != nil {
			return fmt.Errorf("failed to delete the image with uuid: %s, taskUUID: %s: %w", imgUUID, taskUUID, err)
		}
		return nil
	} 
	return fmt.Errorf("failed to convert the task UUID %v to string", respd.Status.ExecutionContext.TaskUUID)
```

File: pkg/infrastructure/nutanix/clusterapi/clusterapi.go
Comment (r4f4) [inline pkg/infrastructure/nutanix/clusterapi/clusterapi.go:275]:
```suggestion

```

File: pkg/infrastructure/nutanix/clusterapi/clusterapi.go
Comment (r4f4) [inline pkg/infrastructure/nutanix/clusterapi/clusterapi.go:152]: Since you're making API calls to upload/delete content, what about using https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait#ExponentialBackoffWithContext which will sleep in between attempts and will stop the operation if the context is cancelled or times out?

File: pkg/infrastructure/nutanix/clusterapi/clusterapi.go
Comment (r4f4) [inline pkg/infrastructure/nutanix/clusterapi/clusterapi.go:174]: Instead of (create or delete) + (upload or delete) in a loop, wouldn't it be better to separate into 2 retriables? That way you just create (and delete, in the failure path) the image once.
1. A loop (wait.ExponentialBackoffWithContext) just to create the image. Return failure if it didn't succeed
2. A loop (wait.ExponentialBackoffWithContext) just to upload the image. Delete image and return failure if it didn't succeed.

File: pkg/infrastructure/nutanix/clusterapi/clusterapi.go
Comment (yanhua121) [reply pkg/infrastructure/nutanix/clusterapi/clusterapi.go]: changed.

File: pkg/infrastructure/nutanix/clusterapi/clusterapi.go
Comment (yanhua121) [reply pkg/infrastructure/nutanix/clusterapi/clusterapi.go]: Changed to use wait#ExponentialBackoffWithContext

File: pkg/infrastructure/nutanix/clusterapi/clusterapi.go
Comment (yanhua121) [reply pkg/infrastructure/nutanix/clusterapi/clusterapi.go]: Made the change as suggested.

File: pkg/infrastructure/nutanix/clusterapi/clusterapi.go
Comment (yanhua121) [reply pkg/infrastructure/nutanix/clusterapi/clusterapi.go]: changed

File: pkg/infrastructure/nutanix/clusterapi/clusterapi.go
Comment (r4f4) [inline pkg/infrastructure/nutanix/clusterapi/clusterapi.go:162]: Notice that returning an error makes the wait abort immediately and return the error. So you might want to consider which errors are retriable and `return false, nil` in those cases.

Comment (yanhua121) [reply pkg/infrastructure/nutanix/clusterapi/clusterapi.go]: fixed.

**Comment by yanhua121** [comment]
@r4f4 All your comments are addressed. Can you take another look? Thanks.

File: pkg/infrastructure/nutanix/clusterapi/clusterapi.go
Comment (r4f4) [inline pkg/infrastructure/nutanix/clusterapi/clusterapi.go:215]: I think the `deleteImage` should be here, not inside the wait/retry loop: you try to upload the image a few times and, if it doesn't work, then you delete the image. Otherwise you're trying to upload an image that has been deleted.

File: pkg/infrastructure/nutanix/clusterapi/clusterapi.go
Comment (r4f4) [inline pkg/infrastructure/nutanix/clusterapi/clusterapi.go:203]: Nit: you can use the `` avoid the escaping:
```suggestion
			if strings.Contains(e1.Error(), `"code": 401`) || strings.Contains(e1.Error(), `"code": 403`) {
```
Here it doesn't much difference but it can be useful for longer strings with lots of escaping.

File: pkg/infrastructure/nutanix/clusterapi/clusterapi.go
Comment (r4f4) [inline pkg/infrastructure/nutanix/clusterapi/clusterapi.go:304]:
```suggestion
		// Wait till the image deletion task is successed.
```

Comment (yanhua121) [reply pkg/infrastructure/nutanix/clusterapi/clusterapi.go]: changed.

**Comment by yanhua121** [comment]
@r4f4 
Please take another look. The code is refactored to put both createImage() and uploadImageData() in the same wait.ExponentialBackoffWithContext(). Because if createImage() succeeds but uploadImage fails, we need to delete the image object and retry to call both createImage() and uploadImageData() again. The old-version prism-api server sometime returns error for the uploadImage call and does not allow to retry the uploadImage() call to the same image object.

**Comment by yanhua121** [comment]
> > @r4f4
> > Please take another look. The code is refactored to put both createImage() and uploadImageData() in the same wait.ExponentialBackoffWithContext(). Because if createImage() succeeds but uploadImage fails, we need to delete the image object and retry to call both createImage() and uploadImageData() again. The old-version prism-api server sometime returns error for the uploadImage call and does not allow to retry the uploadImage() call to the same image object.
> 
> Ah I see. I didn't know that so I misdirected you with my suggestion to separate the code. It would be nice to add this explanation as a comment in the code before the call to delete.

@r4f4 Added this explanation comment before the wait.ExponentialBackoffWithContext() block.

**Comment by r4f4** [comment]
Awesome. I added the approve. I'll leave it up to someone in the nutanix team to add the lgtm.

**Review by thunderboltsid — APPROVED** [review]
/lgtm

**Comment by yanhua121** [comment]
@r4f4 
Nutanix team's @thunderboltsid reviewed and "lgtm"-ed the PR. But he does not have authority to put the "lgtm" label. Can you put the label for the PR to merge? Thanks.

**Pull request merged**
